API endpoints to manage staff permissions

ZacWarham opened the pull request:
Adds `/insert`, `/update`, and `/delete` API endpoints to give the ability to manage staff permissions in a course instance.

<details>
<summary>Example usage</summary>

```bash
curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)", "course_role": "Previewer"}' http://localhost:3000/pl/api/v1/course/1/staff/insert 

curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)", "course_role": "Editor"}' http://localhost:3000/pl/api/v1/course/1/staff/update/

curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)", "course_instance_role": "Student Data Viewer"}' http://localhost:3000/pl/api/v1/course/1/staff/insert/3

curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)", "course_instance_role": "Student Data Editor"}' http://localhost:3000/pl/api/v1/course/1/staff/update/3

curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)"}' http://localhost:3000/pl/api/v1/course/1/staff/delete/3 

curl -X POST -H "Content-Type: application/json" -H "Private-Token: TOKEN" -d '{"uid": "[viewer@example.com](mailto:viewer@example.com)"}' http://localhost:3000/pl/api/v1/course/1/staff/delete
```

</details>

Couple todo items but would be helpful for feedback first (both here or for the questions in the linked issue) to avoid duplicating the work effort
- documentation
- test cases

Closes #12444

nwalters512 reviewed (commented):
Notes for the reviewer:

- `/insert`/`/update`/`/delete` don't seem like conventional REST endpoint to me. There are likely better options here.
- These endpoints are missing in the API documentation.
- These are missing tests.
- Please check very very carefully that permissions checks are being done correctly; tests would likely help with this.

ZacWarham commented:
I have now changed the endpoint paths, added tests, and updated the documentation.

I am not certain on how to handle the permission checks at this stage so feedback here would be appreciated.

ZacWarham commented on docs/api.md:
````diff
@@ -135,3 +149,71 @@ GET /pl/api/v1/course/:course_id/sync/:job_sequence_id
+
+```text
+DELETE /pl/api/v1/course/:course_id/staff/:course_instance_id
+```
````
37bced7

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,210 @@
+import { selectOptionalUserByUid } from '../../../../models/user.js';
+import { CourseUsersRowSchema } from '../../../../pages/instructorCourseAdminStaff/instructorCourseAdminStaff.html.js';
+
+const CourseUsersRowSchemaAPIFriendly = CourseUsersRowSchema.pick({
+  user: true,
+  course_permission: true,
+  course_instance_roles: true,
+});
```
Is there a better way to extract this information?

mylesw27 replied:
This is a fine way to do this. However, I think we may want to update this so we are parsing this for only the information we really want to include here. For example, there are some items in `user` (like `stripe_customer_id`) that we probably do not want to include. 

Also, since this schema is only being used once, we could probably just inline it directly where we call the query.

ZacWarham replied:
Updated in 63d8b1f

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,210 @@
+  course_instance_roles: true,
+});
+
+const sql = sqldb.loadSql(
+  path.join(
+    import.meta.dirname,
+    '../../../..',
+    'pages/instructorCourseAdminStaff/instructorCourseAdminStaff.sql',
+  ),
+);
```
Is it safe to pulling the SQL from here or should I have a new SQL file here with just the required information (instead of filtering from above)?

mylesw27 replied:
I believe this is safe as long as we are using a correct schema for our query. However, it is not really really conventional to reach out to the `pages` directory to grab a query. Consider moving the SQL query and calls to the `models/course-permissions` file that we are already importing from.

ZacWarham replied:
Would you prefer I move the query from the pages directory and then reference to models from both locations it is currently used or duplicate it to models and just reference the API code to models?

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,210 @@
+  asyncHandler(async (req, res) => {
+    const { uid, course_role } = req.body;
+
+    // User can be added to staff list before being created so do not check for user existence
```
Is this comment needed?

mylesw27 replied:
I don't understand the comment. Isn't the following statement doing the opposite?

ZacWarham replied:
Sorry, poorly worded comment. It's meant to be in reference to the PL database since a user can be added to a course as staff if they don't exist so I don't use `validateUid` as opposed to whether the requester has included one in the request body

ZacWarham commented on apps/prairielearn/src/api/v1/index.ts:
```diff
@@ -90,6 +90,10 @@ router.use(
   (await import('./endpoints/courseInstanceAccessRules/index.js')).default,
 );
+router.use(
+  '/course/:course_id(\\d+)/staff',
```
Are `/staff` and `/course_instance/:course_instance(\\d+)` sufficient? 

I think the first is okay but the second seems messy

mylesw27 replied:
These endpoints seem good to me.

ZacWarham commented on apps/prairielearn/src/lib/db-types.ts:
```diff
@@ -26,6 +26,16 @@ export const JsonCommentSchema = z.union([z.string(), z.array(z.any()), z.record
 // Enum schemas. These should be alphabetized by their corresponding enum name.
 // *******************************************************************************
 
+export const EnumCourseInstancePermissionSchema = z.enum([
+  'None',
+  'Student Data Viewer',
+  'Student Data Editor',
+]);
+export type EnumCourseInstancePermission = z.infer<typeof EnumCourseInstancePermissionSchema>;
+
+export const EnumCourseRoleSchema = z.enum(['None', 'Previewer', 'Viewer', 'Editor', 'Owner']);
+export type EnumCourseRole = z.infer<typeof EnumCourseRoleSchema>;
+
```
First time with enums, not sure what the naming conventions should be

mylesw27 replied:
These names look good to me!

ZacWarham commented on apps/prairielearn/src/tests/api.test.ts:
```diff
@@ -383,5 +383,339 @@ describe('API', { timeout: 60_000 }, function () {
+      assert.equal(res.status, 400);
+    });
+
+    test.sequential(
+      'Use GET endpoint to verify that student data access was removed',
+      async function () {
+        const res = await fetch(locals.apiCourseStaffUrl, {
+          headers: {
+            'Private-Token': locals.api_token,
+          },
+        });
+        assert.equal(res.status, 200);
+
+        const json = (await res.json()) as any;
+        assert.exists(json.users);
+        assert.equal(json.users[0].user.uid, 'dev@example.com');
+        assert.equal(json.users[0].course_permission.course_role, 'Owner');
+        assert.isNull(json.users[0].course_instance_roles);
+      },
+    );
```
Is it worth having this as a separate test or alternatively is there a way of including this inside of the post request?

mylesw27 replied:
I think a separate test for this is fine. This way it is testing both the `post` route and the `get` route to verify.

ZacWarham commented on docs/api.md:
````diff
@@ -135,3 +149,71 @@ GET /pl/api/v1/course/:course_id/sync/:job_sequence_id
+POST /pl/api/v1/course/:course_id/staff
+```
+
+```json
+-d '{"uid": "dev@example.com", "course_role": "Previewer"}'
+```
````
1. Do these need to be `json` or should they be `text` or `sh` instead?
2. Is this formatting sufficient?

nwalters512 replied:
Given that `-d '...` is not valid JSON, this should change. There's no need/reason to use curl-specific things here. I'd just document this like this:

> ##### Example body
>
> ```json
> {"uid": "dev@example.com", "course_role": "Previewer"}
> ```

ZacWarham replied:
Sorry for all the commits fixing this, my PC is broken so I did this in the browser and made a few mistakes

8b2cc6f
051c768
17c87f2

ZacWarham reviewed (commented):
Just did a self-review with some questions for when you have a chance @mylesw27 

Should be good to go on my end otherwise once I go through any feedback you have

mylesw27 reviewed (commented):
Tested locally and the endpoints are working correctly. I agree that there needs to be better permissions testing. I'm not entirely sure if we need to create a second `api_token` for a different user (that does not have the correct permissions) or make a second class in the tests where the default user does not have the correct permissions. @nwalters512 Would you be able to guide us (or link to some examples) on how we would be able to test permissions for this?

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,217 @@
+import {
+  EnumCourseInstancePermissionSchema,
+  EnumCourseRoleSchema,
+} from '../../../../lib/db-types.js';
```
Fixed in 3027c54

ZacWarham commented:
Updates from the review have been added or requested for more information

Re. some of the code rabbits comments
- https://github.com/PrairieLearn/PrairieLearn/pull/12474#discussion_r2283880816
- https://github.com/PrairieLearn/PrairieLearn/pull/12474#discussion_r2283880822

Is it okay to assume that given the test suite should only be creating the lists of these sizes that this is not needed?

nwalters512 commented:
> Is it okay to assume that given the test suite should only be creating the lists of these sizes that this is not needed?

This is probably reasonable. The tests should still fail regardless if at some point that assumption stops being true.

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,218 @@
+  const validate_role = EnumCourseRoleSchema.safeParse(course_role);
+
+  if (!validate_role.success) {
+    throw new error.HttpStatusError(404, 'Role not found');
```
Switched in [7ab8681](https://github.com/PrairieLearn/PrairieLearn/pull/12474/commits/7ab86819c099b04cf48c9e79dafb6456ca878846)

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,218 @@
+
+    validateCourseRole({ course_role });
+
+    const job = await insertCoursePermissionsByUserUid({
```
Mistake on my end - renamed to user c622a01

ZacWarham commented on apps/prairielearn/src/tests/api.test.ts:
```diff
@@ -383,5 +391,379 @@ describe('API', { timeout: 60_000 }, function () {
+        name,
+        token,
+        token_hash,
+      });
```
I added this in for here and `userSettings` (e0445bf) but I am not confident that it is best practice to sometimes return the token the way it is now?

ZacWarham replied:
I do not recall what this concern was now. I believe my question was "is it safe to return the token instead of just creating it and returning nothing" but it is required at times so not sure

ZacWarham commented:
Note to self for today's progress

> I added this in for here and `userSettings` ([e0445bf](https://github.com/PrairieLearn/PrairieLearn/commit/e0445bfe712fd2cd5f800d9d678a7c174ade3c2e)) but I am not confident that it is best practice to sometimes return the token the way it is now?

The function inputs should be typed

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,218 @@
+      course_role,
+      authn_user_id: res.locals.authz_data.authn_user.user_id,
+    });
+    res.status(200).json({ user: job });
```
Updated in 8bd96cc

ZacWarham commented on apps/prairielearn/src/api/v1/endpoints/courseStaffPermissions/index.ts:
```diff
@@ -0,0 +1,218 @@
+);
+
+// Update user access to student data
+router.put(
```
@nwalters512 In implementing this combined route, it appears that both insert functions work as updates. Is it safe to use to use these for both insert and update or should I determine if the user already exists and use the `update` equivalent function if so?

https://github.com/PrairieLearn/PrairieLearn/blob/063992f80e36247ff07796cc07025fde96048a21/apps/prairielearn/src/models/course-permissions.ts#L127

https://github.com/PrairieLearn/PrairieLearn/blob/063992f80e36247ff07796cc07025fde96048a21/apps/prairielearn/src/models/course-permissions.ts#L24
